nodes/hybrid-consensus/src/service.rs

```diff
@@ -0,0 +1,247 @@
+//! Service and ServiceFactory implementation. Specialized wrapper over substrate service.
+
+use std::sync::Arc;
+use std::time::Duration;
+use sc_client::LongestChain;
+use sc_client_api::ExecutorProvider;
+use runtime::{self, opaque::Block, RuntimeApi};
+use sc_service::{error::{Error as ServiceError}, AbstractService, Configuration, ServiceBuilder};
+use sp_inherents::InherentDataProviders;
+use sc_executor::native_executor_instance;
+pub use sc_executor::NativeExecutor;
+use sha3pow::MinimalSha3Algorithm;
+use sc_finality_grandpa::{
+	self,
+	FinalityProofProvider as GrandpaFinalityProofProvider,
+	StorageAndProofProvider
+};
+
+// Our native executor instance.
+native_executor_instance!(
+	pub Executor,
+	runtime::api::dispatch,
+	runtime::native_version,
+);
+
+pub fn build_inherent_data_providers() -> Result<InherentDataProviders, ServiceError> {
+	let providers = InherentDataProviders::new();
+
+	providers
+		.register_provider(sp_timestamp::InherentDataProvider)
+		.map_err(Into::into)
+		.map_err(sp_consensus::error::Error::InherentData)?;
+
+	Ok(providers)
+}
+
+/// Starts a `ServiceBuilder` for a full service.
+///
+/// Use this macro if you don't actually need the full service, but just the builder in order to
+/// be able to perform chain operations.
+macro_rules! new_full_start {
+	($config:expr) => {{
+		let mut import_setup = None;
+		let inherent_data_providers = crate::service::build_inherent_data_providers()
+			.expect("Inherent data providers are present.");
+
+		let builder = sc_service::ServiceBuilder::new_full::<
+			runtime::opaque::Block, runtime::RuntimeApi, crate::service::Executor
+		>($config)?
+			.with_select_chain(|_config, backend| {
+				Ok(sc_client::LongestChain::new(backend.clone()))
+			})?
+			.with_transaction_pool(|config, client, _fetcher| {
+				let pool_api = sc_transaction_pool::FullChainApi::new(client.clone());
+				Ok(sc_transaction_pool::BasicPool::new(config, std::sync::Arc::new(pool_api)))
+			})?
+			.with_import_queue(|_config, client, mut select_chain, _transaction_pool| {
+				let select_chain = select_chain.take()
+					.ok_or_else(|| sc_service::Error::SelectChainRequired)?;
+				let (grandpa_block_import, grandpa_link) =
+					sc_finality_grandpa::block_import(
+						client.clone(), &(client.clone() as std::sync::Arc<_>), select_chain.clone()
+					)?;
+
+				//TODO do I need this justification import? Seems to work without it.
+				// let justification_import = grandpa_block_import.clone();
```

Comment:
This is used when importing justifications through the sync module. Justifications will be requested from peers by the sync module whenever we import an authority set change block that doesn't contain a justification.

Reply:
Can you drop a little more knowledge here? What is a justification? Is it a proof that a block as been finalized by grandpa? Or something else?

Do I need to create one and do something with it in this case? If yes, what should I do? If no, why is this different than the "standard" babe/grandpa situation?

Reply:
BABE/GRANDPA is a hybrid consensus mechanism where both systems share the same authority set. The authority set itself is "managed" by BABE, in the sense that is defined in the runtime. Whenever the authority set changes an event is added to the block (https://github.com/paritytech/substrate/blob/master/frame/grandpa/src/lib.rs#L208). From the POV of BABE these authority set changes are immediate (i.e. after that block the authority set has changed), for GRANDPA we can't do that otherwise we'd have insecure authority handoffs (what would happen if we changed it immediately and then there was a re-org?). So instead GRANDPA only triggers these authority set changes when the block that signaled them has been finalized (this is why authority sets can be unsynchronized between BABE and GRANDPA, the GRANDPA authority set can be lagging a bit compared to BABE's). When importing a block that triggers authority set changes the GRANDPA voter will start limiting its votes to that block (to make sure that it gets finalized), afterwards once this block is finalized we generate a justification which justifies why that block is finalized (it is composed of a GRANDPA commit and the set of headers required to interpret that commit https://github.com/paritytech/substrate/blob/master/client/finality-grandpa/src/justification.rs#L40). This justification is then stored in the blocks database so that it can be presented to any later syncing client.

When syncing the chain, after importing a block that triggers authority set changes, the client will either verify that a justification is attached to the block (and verify it), or in case there is no justification attached to the block (maybe because the block hasn't been finalized yet) the client will ask the sync protocol to keep trying to fetch this justification from its peers. Only after this justification is imported will the block get finalized which will trigger the actual authority set change for GRANDPA. Since the sync protocol is agnostic to the actual import logic it needs to have some handle to how to import these justifications after it gets them (just like for block import).

So, to answer your question, you need to pass `Some(grandpa_block_import)`, which is the handle that sync logic will use to import justifications if it ever needs them. The justifications will be requested automatically and you don't need to do anything else as this logic is implemented in `GrandpaBlockImport`.